modules/mono/glue/GodotSharp/GodotSharp/Core/Array.cs

```diff
@@ -273,36 +480,207 @@ public void Add(Variant item)
+        public void AddRange(IEnumerable<Variant> collection)
+        {
+            if (collection == null)
+                throw new ArgumentNullException(nameof(collection), "Value cannot be null.");
+
+            // If the collection is another Godot Array, we can add the items
+            // with a single interop call.
+            if (collection is Array array)
+            {
+                var self = (godot_array)NativeValue;
+                var collectionNative = (godot_array)array.NativeValue;
+                _ = NativeFuncs.godotsharp_array_add_range(ref self, collectionNative);
+                return;
+            }
+            if (collection is IGenericGodotArray typedArray)
+            {
+                var self = (godot_array)NativeValue;
+                var collectionNative = (godot_array)typedArray.UnderlyingArray.NativeValue;
+                _ = NativeFuncs.godotsharp_array_add_range(ref self, collectionNative);
+                return;
+            }
+
+            // If we can retrieve the count of the collection without enumerating it
+            // (e.g.: the collections is a List<T>), use it to resize the array once
+            // instead of growing it as we add items.
+            if (collection.TryGetNonEnumeratedCount(out int count))
+            {
+                Resize(Count + count);
+
+                var enumerator = collection.GetEnumerator();
+
+                for (int i = 0; i < count; i++)
+                {
+                    enumerator.MoveNext();
+                    this[count + i] = enumerator.Current;
+                }
+
+                return;
+            }
+
+            foreach (var item in collection)
+            {
+                Add(item);
+            }
+        }
+
+        /// <summary>
+        /// Finds the index of an existing value (or the insertion index that
+        /// maintains sorting order, if the value is not yet present in the array)
+        /// using binary search. Optionally, a <paramref name="before"/> specifier
+        /// can be passed. If <see langword="false"/>, the returned index comes
+        /// after all existing entries of the value in the array.
+        /// Note: Calling <see cref="BinarySearch"/> on an unsorted array results
+        /// in unexpected behavior.
+        /// </summary>
+        /// <param name="item">The <see cref="Variant"/> item to search for.</param>
+        /// <param name="before">
+        /// If <see langword="false"/>, the returned index comes after all existing
+        /// entries of the value in the array.
+        /// </param>
+        /// <returns>
+        /// The index of the found value or the insertion index that maintains
+        /// sorting order.
+        /// </returns>
+        public int BinarySearch(Variant item, bool before = true)
+        {
+            if (Count == 0)
+            {
+                // Special case for empty array to avoid an interop call.
+                return -1;
```

Comment:
```suggestion
                return 0;
```
Returns 0 for empty arrays.